dot/state/pruner/pruner.go

```diff
@@ -0,0 +1,354 @@
+package pruner
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/scale"
+	log "github.com/ChainSafe/log15"
+)
+
+const (
+	journalPrefix = "journal"
+	lastPrunedKey = "last_pruned"
+	pruneInterval = time.Second
+)
+
+// nolint
+const (
+	Full    = Mode("full")
+	Archive = Mode("archive")
+)
+
+// Mode online pruning mode of historical state tries
+type Mode string
+
+// IsValid checks whether the pruning mode is valid
+func (p Mode) IsValid() bool {
+	switch p {
+	case Full:
+		return true
+	case Archive:
+		return true
+	default:
+		return false
+	}
+}
+
+// Pruner is implemented by FullNode and ArchiveNode.
+type Pruner interface {
+	StoreJournalRecord(deleted, inserted []common.Hash, blockHash common.Hash, blockNum int64) error
+}
+
+// ArchiveNode is a no-op since we don't prune nodes in archive mode.
+type ArchiveNode struct{}
+
+// StoreJournalRecord for archive node doesn't do anything.
+func (a *ArchiveNode) StoreJournalRecord(deleted, inserted []common.Hash, blockHash common.Hash, blockNum int64) error {
+	return nil
+}
+
+type deathRecord struct {
+	blockHash   common.Hash
+	deletedKeys map[common.Hash]int64 // Mapping from deleted key hash to block number.
+}
+
+type deathRow []*deathRecord
+
+// FullNode stores state trie diff and allows online state trie pruning
+type FullNode struct {
+	logger        log.Logger
+	deathList     []deathRow
+	storageDB     chaindb.Database
+	journalDB     chaindb.Database
+	deathIndex    map[common.Hash]int64 // Mapping from deleted key hash to block number.
+	pendingNumber int64                 // block number to be pruned. Initial value is set to 1 and is incremented after every block pruning.
+	retainBlocks  int64
+	sync.RWMutex
+}
+
+type journalRecord struct {
+	// blockHash of the block corresponding to journal record
+	blockHash common.Hash
+	// Hash of keys that are inserted into state trie of the block
+	insertedKeys []common.Hash
+	// Hash of keys that are deleted from state trie of the block
+	deletedKeys []common.Hash
+}
+
+type journalKey struct {
+	blockNum  int64
+	blockHash common.Hash
+}
+
+func newJournalRecord(hash common.Hash, insertedKeys, deletedKeys []common.Hash) *journalRecord {
+	return &journalRecord{
+		blockHash:    hash,
+		insertedKeys: insertedKeys,
+		deletedKeys:  deletedKeys,
+	}
+}
+
+// NewFullNode creates a Pruner for full node.
+func NewFullNode(db, storageDB chaindb.Database, retainBlocks int64, l log.Logger) (Pruner, error) {
+	p := &FullNode{
+		deathList:    make([]deathRow, 0),
+		deathIndex:   make(map[common.Hash]int64),
+		storageDB:    storageDB,
+		journalDB:    chaindb.NewTable(db, journalPrefix),
+		retainBlocks: retainBlocks,
+		logger:       l,
+	}
+
+	blockNum, err := p.getLastPrunedIndex()
+	if err != nil {
+		return nil, err
+	}
+
+	p.logger.Info("last pruned block", "block num", blockNum)
```

Comment:
change to debug

Reply:
This log is printed only once during the entire runtime and indicates that the pruner is being initialized.

Reply:
yes but I think you can still change this to debug